File: microservice-architecture/api-versioning/go/src/router.go
// Package src 实现了 API 版本管理模式（API Versioning Pattern）的核心逻辑。
//
// 支持 URL 路径版本（/v1/、/v2/）和 Header 版本（X-API-Version），提供默认版本兜底。
//
// 【设计模式】
//   - 策略模式（Strategy Pattern）：不同版本的处理器是可互换的策略，
//     路由器根据解析到的版本号选择对应策略执行。
//   - 工厂方法模式（Factory Method）：ResolveVersion 根据请求上下文决定使用哪个版本。
//   - 模板方法模式（Template Method）：Handle 定义了"解析版本 → 查找处理器 → 执行"的骨架。
//
// 【架构思想】
//   API 版本管理让新老客户端并行使用不同版本接口，实现平滑演进。
//
// 【开源对比】
//   - go-chi / gorilla/mux：通过 URL 路径参数实现版本路由
//   - grpc-go：通过 package 名或 service 名区分版本
//   本示例省略了版本协商和废弃通知等工程细节，聚焦于版本解析和路由分发。
package src

import "strings"

type Request struct {
	Path    string
	Headers map[string]string
}

type Response struct {
	StatusCode int
	Version    string
	Body       string
}

type VersionedRouter struct {
	defaultVersion string
	handlers       map[string]func() string
}

func NewVersionedRouter(defaultVersion string) *VersionedRouter {
	return &VersionedRouter{
		defaultVersion: defaultVersion,
		handlers:       map[string]func() string{},
	}
}

func (r *VersionedRouter) Register(version string, handler func() string) {
	r.handlers[normalizeVersion(version)] = handler
}

func (r *VersionedRouter) Handle(req Request) Response {
	version := r.ResolveVersion(req)
	handler := r.handlers[version]
	if handler == nil {
		return Response{
			StatusCode: 400,
			Version:    version,
			Body:       "unsupported api version",
		}
	}

	return Response{
		StatusCode: 200,
		Version:    version,
		Body:       handler(),
	}
}

func (r *VersionedRouter) ResolveVersion(req Request) string {
	path := strings.ToLower(req.Path)
	if strings.Contains(path, "/v2/") {
		return "v2"
	}
	if strings.Contains(path, "/v1/") {
		return "v1"
	}

	headerVersion := normalizeVersion(req.Headers["X-API-Version"])
	if headerVersion != "" {
		return headerVersion
	}

	return normalizeVersion(r.defaultVersion)
}

func ProductHandlerV1() string {
	return `{"id":"P100","name":"Mechanical Keyboard"}`
}

func ProductHandlerV2() string {
	return `{"id":"P100","name":"Mechanical Keyboard","inventoryStatus":"IN_STOCK"}`
}

func normalizeVersion(version string) string {
	version = strings.TrimSpace(strings.ToLower(version))
	if version == "" {
		return ""
	}
	if strings.HasPrefix(version, "v") {
		return version
	}
	return "v" + version
}

File: microservice-architecture/api-versioning/c/src/router.c
/**
 * @file router.c - API 版本管理模式（API Versioning Pattern）的 C 语言实现
 *
 * 实现 URL 路径版本解析和 Header 版本回退逻辑。
 */

#include "func.h"

#include <stdio.h>
#include <string.h>

static const char *normalize_version(const char *version)
{
    if (version == NULL || version[0] == '\0') {
        return "";
    }
    return version[0] == 'v' ? version : (strcmp(version, "2") == 0 ? "v2" : "v1");
}

void version_router_handle(const VersionRequest *request, VersionResponse *response)
{
    const char *version = "";
    if (strstr(request->path, "/v2/") != NULL) {
        version = "v2";
    } else if (strstr(request->path, "/v1/") != NULL) {
        version = "v1";
    } else if (request->api_version[0] != '\0') {
        version = normalize_version(request->api_version);
    } else {
        version = "v1";
    }

    strcpy(response->version, version);
    if (strcmp(version, "v1") == 0) {
        response->status_code = 200;
        strcpy(response->body, "{\"id\":\"P100\",\"name\":\"Mechanical Keyboard\"}");
    } else if (strcmp(version, "v2") == 0) {
        response->status_code = 200;
        strcpy(response->body, "{\"id\":\"P100\",\"name\":\"Mechanical Keyboard\",\"inventoryStatus\":\"IN_STOCK\"}");
    } else {
        response->status_code = 400;
        strcpy(response->body, "unsupported api version");
    }
}
